### PR TITLE
refactor: Add fine grained (object level) catalog locking

### PR DESCRIPTION
**Pull request by alamb**
This PR is the first phase of https://github.com/influxdata/influxdb_iox/issues/1042

# Rationale:
The eventual goal is to hang `DBChunk` instances off  `catalog::Chunk` instances so that there is no possibility of inconsistency.

In order to allow writes to open chunks attached to such chunks, I need to be able to prevent their "state" from changing (aka so that we don't start moving data out into the read buffer) until all outstanding writes to them have completed. This actually may be a hole in the current code

I was hoping (for reasons I can't articulate) to avoid this pattern of direct manipulation of catalog objects, but since it seems to be required to combine all the chunks together I just went with it. 

# Changes
1. Collapse the catalog out of its own crate and into the `server` crate
2. Add "fine grained locking" on chunks and partitions

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: server/src/db.rs [hunk 1]
```diff
@@ -212,41 +214,47 @@ impl Db {
             .as_ref()
             .context(DatatbaseNotWriteable)?;
 
-        // hold catalog lock to prevent any concurrent modifications
-        let mut catalog = self.catalog.write();
```
Comment (alamb) [inline server/src/db.rs:216]: The crux of the changes are rather than locking the entire catalog, the code now locks individual `Partition`s and `Chunk`s

**Comment by mkmik** [comment]
> Collapse the catalog out of its own crate and into the server crate
why?

**Comment by mkmik** [comment]
> > Collapse the catalog out of its own crate and into the server crate
> 
> why?

I mean, I'm just curious; I'm trying to figure out when it makes sense to create a new crate/package; I'm writing some new code and I moved it into its own crate and it compiles so much faster. I was wondering I would be merging this new code into an existing crate or keep it in it's own crate; now that I see that you ended up merging it back, I wonder what should I do.

**Comment by alamb** [comment]
> I mean, I'm just curious; I'm trying to figure out when it makes sense to create a new crate/package; I'm writing some new code and I moved it into its own crate and it compiles so much faster. I was wondering I would be merging this new code into an existing crate or keep it in it's own crate; now that I see that you ended up merging it back, I wonder what should I do.

@mkmik  I personally like to keep things in separate crates (mostly so I can run the tests individually (`cargo test -p <cratename>`) which is typically quite a bit faster

In this case I brought the code into the `server` crate because I plan to begin storing objects from other crates (mutable buffer, etc) and so the boundaries of what is stored in `Catalog` and what is stored in the rest of the `Db` struct is going to get much less crisp

File: server/src/db/catalog.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    #[snafu(display("chunk already exists: {}:{}", partition_key, chunk_id))]
+    ChunkAlreadyExists {
+        partition_key: String,
+        chunk_id: u32,
+    },
+}
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub mod chunk;
```
Comment (e-dard) [inline server/src/db/catalog.rs:32]: Could these all be moved up top to tidy things up a bit?

**Review by e-dard — APPROVED** [review]
LGTM 👍

File: server/src/db/catalog.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+///
+/// The Catalog stores information such as which chunks exist, what
+/// state they are in, and what objects on object store are used, etc.
+///
+/// The catalog is also responsible for (eventually) persisting this
+/// information
+#[derive(Default, Debug)]
+pub struct Catalog {
+    /// key is partition_key
+    partitions: RwLock<BTreeMap<String, Arc<RwLock<Partition>>>>,
```
Comment (tustvold) [inline server/src/db/catalog.rs:47]: `Arc<RwLock>` makes me a bit apprehensive. It makes ownership a little funny, but more importantly it allows holding a read/write lock on a partition without a read lock on the catalog - which may not be a good thing?

**Comment by tustvold** [comment]
Having had a play around I'm not sure I understand the motivation for using Arcs.

If the catalog-level RwLock is kept outside the Catalog object itself (as before), I think it removes the only need to support Arc that I can see in this code.

There is likely going to need to be some level of Arcs within the chunks themselves, as part of query-"snapshotting", but unless I'm mistaken that should be of immutable and therefore mutexless data?

**Comment by alamb** [comment]
> Having had a play around I'm not sure I understand the motivation for using Arcs.
>
> If the catalog-level RwLock is kept outside the Catalog object itself (as before), I think it removes the only need to support Arc that I can see in this code, and avoids potentially surprising issues where it is possible to mutate a partition without a read-lock on the catalog itself

The motivation / rationale for using Arcs is to enable fine grained locking on the partitions and chunks rather than the entire catalog. 

With this PR there is no way to lock the entire catalog -- there is a lock to keep the `partitions` list consistent but the locks happen on individual partitions. 

Perhaps you can sketch out an alternate approach without Arcs? There may well be a better way to do fine grained locking, but it isn't clear to me what it is

**Comment by tustvold** [comment]
Had a good discussion with @alamb and it ultimately comes down to a tradeoff between maximal concurrency and the ability for the Rust compiler to prevent mistakes. This PR favors concurrency over compile-time validation, which is not necessarily a bad thing. It does, however, mean extra care will need to be taken to manually take the necessary locks when performing partition or catalog level manipulations, such as computing aggregate statistics.

As currently the catalog is just a "dumb" collection of chunks, the discussion is largely academic, and so I'm happy to move forward with this PR as-is. If this becomes a problem in the future we can always revisit.

**Review by tustvold — APPROVED** [review]

**Comment by alamb** [comment]
Thanks for the clarification @tustvold  -- will polish this PR up and get it in

**Pull request merged**